src/basic/time-util.c

```diff
@@ -910,14 +908,26 @@ int parse_timestamp(const char *t, usec_t *usec) {
         }
 
         if (pid == 0) {
+                bool with_tz = true;
+
                 if (setenv("TZ", t_timezone, 1) != 0) {
                         shared->return_value = negative_errno();
                         _exit(EXIT_FAILURE);
                 }
 
                 tzset();
 
-                shared->return_value = parse_timestamp_impl(t, &shared->usec, true);
+                /* If there is a timezone that matches the tzname fields, leave the parsing to the implementation
+                 * Otherwise just cut it off */
+                with_tz = !STR_IN_SET(t_timezone, tzname[0], tzname[1]);
+
+                /*cut of the timezone if we dont need it*/
+                if (t_timezone && !with_tz)
+                        t = strdup(t);
+                else
+                        t = strndupa(t, last_space - t);
```

Comment:
this is really strange, in one case you allocate from heap and in the other case from stack... (strdup() allocates from heap, but strndupa() from stack -- the `a` suffix indicates that)